### PR TITLE
Ft 12 composition tokens 473

### PR DESCRIPTION
**Pull request by swordEdge**

### CODE REVIEW

**Review by six7 — CHANGES_REQUESTED** [review]
Bit of functional feedback:
- Can you use a Dropdown for the Property Picker? Like the Selector that appears when you Right click a token or when you click bottom left of the plugin? This would have better support and in this case we don't need the user to type in the property
image

- I currently get an error when I try to open the Inspect panel and a composition token with 2 properties is applied 
image

**Comment by swordEdge** [comment]
> * Can you use a Dropdown for the Property Picker? Like the Selector that appears when you Right click a token or when you click bottom left of the plugin? This would have better support and in this case we don't need the user to type in the property

In Jira board, It was supposed to use Autocomplete, so, Requirement of this task is changed?

> * I currently get an error when I try to open the Inspect panel and a composition token with 2 properties is applied

It's fixed.

**Comment by six7** [comment]
> In Jira board, It was supposed to use Autocomplete, so, Requirement of this task is changed? @six7 

Let's make sure to use the dropdown, yes. 👍

File: src/app/components/CompositionTokenForm.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    if (Array.isArray(internalEditToken.value)) {
+      setValue([...internalEditToken.value, newToken]);
+    } else {
+      setValue([internalEditToken.value, newToken]);
+    }
+  }, [internalEditToken]);
+
+  const removeToken = useCallback((index) => {
+    if (Array.isArray(internalEditToken.value)) {
+      setValue(internalEditToken.value.filter((_, i) => i !== index));
```
Comment (LiamMartens) [inline src/app/components/CompositionTokenForm.tsx:144]: Could we just update this to `setValue((internalEditToken.value as (TokenBoxshadowValue | TokenCompositionValue)[]).filter((_, i) => i !== index));` to not have a typing error

File: src/app/components/EditTokenForm.tsx [hunk 2]
```diff
@@ -114,6 +115,17 @@ function EditTokenForm({ resolvedTokens }: Props) {
     [internalEditToken],
   );
 
+  const handleCompositionChange = React.useCallback(
+    (style: SingleCompositionToken['value']) => {
+      setError(null);
+      if (internalEditToken?.type === TokenTypes.COMPOSITION) {
+        setInternalEditToken(prev => ({ ...prev, value: style }));
```
Comment (LiamMartens) [inline src/app/components/EditTokenForm.tsx:122]: This appears to give a  typing error, can we fix that?

File: src/app/components/PropertySwitchMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
+import * as PropertySwitchMenuPrimitive from '@radix-ui/react-dropdown-menu';
+import { styled, keyframes } from '@/stitches.config';
```
Comment (LiamMartens) [inline src/app/components/PropertySwitchMenu.tsx:2]: `keyframes` can be removed, it's unused

File: src/app/components/SelectableInput.tsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  onChange,
+  data,
+  defaultData,
+}: {
+  name: string;
+  onChange: React.ChangeEventHandler<HTMLSelectElement>;
+  data?: object[];
+  defaultData?: object;
+  value?: string | Number;
+  required: boolean
```
Comment (LiamMartens) [inline src/app/components/SelectableInput.tsx:15]: Can we remove `required` and `value` from the props? It doesn't look like they are being used

File: src/app/components/SelectableInput.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import Select from 'react-select';
+
+export default function SingleSelect({
+  name,
+  onChange,
+  data,
+  defaultData,
+}: {
+  name: string;
+  onChange: React.ChangeEventHandler<HTMLSelectElement>;
```
Comment (LiamMartens) [inline src/app/components/SelectableInput.tsx:11]: This typing is not aligned with `react-select` and is causing an error at the moment, can we fix that as well

File: src/app/components/TokenButton/TokenButton.tsx [hunk 6]
```diff
@@ -115,7 +116,20 @@ export const TokenButton: React.FC<Props> = ({
       [propsToSet[0].name || propsToSet[0]]: propsToSet[0].forcedValue || value,
     };
     if (propsToSet[0].clear) propsToSet[0].clear.map((item) => Object.assign(newProps, { [item]: 'delete' }));
-    setPluginValue(newProps);
+    if (type === 'composition' && value === 'delete') {
+      // distructure composition token when it is unselected
+      const compositionToken = tokensContext.resolvedTokens.find((token) => token.name === tokenValue);
+      let tokensInCompositionToken: NodeTokenRefMap = {};
```
Comment (LiamMartens) [inline src/app/components/TokenButton/TokenButton.tsx:122]: This can be `const`

File: src/app/components/TokenButton/TokenButton.tsx [hunk 7]
```diff
@@ -115,7 +116,20 @@ export const TokenButton: React.FC<Props> = ({
...
+      if (Array.isArray(compositionToken?.rawValue)) {
+        compositionToken?.rawValue.map((token) => {
```
Comment (LiamMartens) [inline src/app/components/TokenButton/TokenButton.tsx:124]: Use `forEach` instead, were not returning anything

File: src/app/components/TokenButton/TokenButton.tsx [hunk 8]
```diff
@@ -115,7 +116,20 @@ export const TokenButton: React.FC<Props> = ({
...
+          tokensInCompositionToken[token.property] = 'delete';
```
Comment (LiamMartens) [inline src/app/components/TokenButton/TokenButton.tsx:125]: These are also causing typing errors but Im not sure we can quickly fix those so leave these for now..

File: src/app/components/TokenTooltip/TokenTooltipContentValue.tsx [hunk 9]
```diff
@@ -1,13 +1,14 @@
 import React from 'react';
 import { useUIDSeed } from 'react-uid';
 import useTokens from '../../store/useTokens';
-import { SingleToken, SingleTypographyToken } from '@/types/tokens';
+import { SingleCompositionToken, SingleToken, SingleTypographyToken } from '@/types/tokens';
```
Comment (LiamMartens) [inline src/app/components/TokenTooltip/TokenTooltipContentValue.tsx:4]: This imported type is unused

File: src/hooks/useGetActiveState.ts [hunk 10]
```diff
@@ -18,11 +17,11 @@ export function useGetActiveState(properties: (PropertyObject | TokenTypes)[], t
 
     return (
       uiState.mainNodeSelectionValues[type] === name
-      || properties.some((prop) => (
+      || properties.some((prop) => {
         // @TODO: This no longer seems to work for color tokens as prop is always string?
```
Comment (LiamMartens) [inline src/hooks/useGetActiveState.ts:21]: This should return that value, so keep the `()`

File: src/plugin/pluginData.ts [hunk 11]
```diff
@@ -12,6 +13,7 @@ import { tokensSharedDataHandler } from './SharedDataHandler';
 import { defaultWorker } from './Worker';
 import { ProgressTracker } from './ProgressTracker';
 import { SelectionGroup, SelectionValue } from '@/types';
+import { TokenCompositionValue } from '@/types/values';
```
Comment (LiamMartens) [inline src/plugin/pluginData.ts:16]: This is unused

File: src/plugin/pluginData.ts [hunk 12]
```diff
@@ -115,7 +114,29 @@ export async function updatePluginData({
   entries.forEach(({ node, tokens }) => {
     promises.add(defaultWorker.schedule(async () => {
       const currentValuesOnNode = tokens ?? {};
-      const newValuesOnNode = { ...currentValuesOnNode, ...values };
+      let newValuesOnNode = {};
```
Comment (LiamMartens) [inline src/plugin/pluginData.ts:117]: Can we assign a type to this? `let newValuesOnNode: NodeTokenRefMap = {};`

**Review by LiamMartens — CHANGES_REQUESTED** [review]

File: src/app/components/CompositionTokenForm.tsx [hunk 13]
```diff
@@ -0,0 +1,186 @@
+import React, { useCallback, useState } from 'react';
+import { CompositionTokenSingleValue } from '@/types/propertyTypes';
```
[the remark this answers is not in the document]
Comment (swordEdge) [reply src/app/components/CompositionTokenForm.tsx:2]: fixed

File: src/app/components/CompositionTokenForm.tsx [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+          fontSize: '$5 !important',
+          '& > div > input': {
+            flex: 2,
+            marginRight: '$5',
+            height: '$10'
+          }
+        }
+      }}
+      >
+        <PropertySwitchMenu open={menuOpened} onOpenChange={() => setMenuOpened(!menuOpened)}>
```
[the remark this answers is not in the document]
Comment (swordEdge) [reply src/app/components/CompositionTokenForm.tsx:77]: fixed

File: src/app/components/CompositionTokenForm.tsx [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+          <PropertySwitchMenuMainTrigger>
+            <span>{token.property}</span>
+          </PropertySwitchMenuMainTrigger>
+          <PropertySwitchMenuContent sideOffset={2}>
+            <PropertySwitchMenuRadioGroup value={token.property}>
+              {properties.length > 0
+                && properties.map((property, index) => (
+                  <PropertySwitchMenuRadioItem value={property} key={index} onSelect={() => onPropertySelected(property)}>
```
[the remark this answers is not in the document]
Comment (swordEdge) [reply src/app/components/CompositionTokenForm.tsx:85]: done

File: src/app/components/CompositionTokenForm.tsx [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+          onChange={onAliasChange}
+          type="text"
+          name="alias"
+          placeholder="Alias"
+        />
+        <Box css={{ width: '$5', marginRight: '$3' }}>
+          <IconButton
+            tooltip="Remove this style"
+            dataCy="button-style-remove-multiple"
+            onClick={() => onRemove(index)}
```
[the remark this answers is not in the document]
Comment (swordEdge) [reply src/app/components/CompositionTokenForm.tsx:107]: done

File: src/app/components/CompositionTokenForm.tsx [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+  PropertySwitchMenu,
+  PropertySwitchMenuContent,
+  PropertySwitchMenuMainTrigger,
+  PropertySwitchMenuRadioGroup,
+  PropertySwitchMenuRadioItem
+} from './PropertySwitchMenu';
+import { Properties } from '@/constants/Properties';
+import { EditTokenObject } from '../store/models/uiState';
+
+function SingleStyleInput({
```
[the remark this answers is not in the document]
Comment (swordEdge) [reply src/app/components/CompositionTokenForm.tsx:19]: done

File: src/app/components/CompositionTokenForm.tsx (hunk 1 above)
Comment (swordEdge) [reply src/app/components/CompositionTokenForm.tsx:144]: There is no typing error already. So I think I don't have to change this.

File: src/app/components/PropertySwitchMenu.tsx (hunk 3 above)
Comment (swordEdge) [reply src/app/components/PropertySwitchMenu.tsx:2]: done

File: src/app/components/SelectableInput.tsx (hunk 4 above)
Comment (swordEdge) [reply src/app/components/SelectableInput.tsx:15]: fixed. I removed SelectableInput.tsx cause It is not used anywhere.

File: src/app/components/SelectableInput.tsx (hunk 5 above)
Comment (swordEdge) [reply src/app/components/SelectableInput.tsx:11]: fixed. I removed SelectableInput.tsx cause It is not used anywhere.

File: src/app/components/TokenButton/TokenButton.tsx (hunk 7 above)
Comment (swordEdge) [reply src/app/components/TokenButton/TokenButton.tsx:124]: fixed

File: src/app/components/TokenButton/TokenButton.tsx (hunk 6 above)
Comment (swordEdge) [reply src/app/components/TokenButton/TokenButton.tsx:122]: fixed

File: src/app/components/TokenTooltip/TokenTooltipContentValue.tsx (hunk 9 above)
Comment (swordEdge) [reply src/app/components/TokenTooltip/TokenTooltipContentValue.tsx:4]: fixed

File: src/hooks/useGetActiveState.ts (hunk 10 above)
Comment (swordEdge) [reply src/hooks/useGetActiveState.ts:21]: fixed

File: src/plugin/pluginData.ts (hunk 11 above)
Comment (swordEdge) [reply src/plugin/pluginData.ts:16]: fixed

File: src/plugin/pluginData.ts (hunk 12 above)
Comment (swordEdge) [reply src/plugin/pluginData.ts:117]: fixed

File: src/app/components/TokenButton/TokenButton.tsx (hunk 8 above)
Comment (swordEdge) [reply src/app/components/TokenButton/TokenButton.tsx]: fixed.

File: src/app/components/CompositionTokenForm.tsx [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+
+export default function CompositionTokenForm({
+  internalEditToken,
+  setValue,
+}: {
+  internalEditToken: EditTokenObject;
+  setValue: (style: CompositionTokenSingleValue | CompositionTokenSingleValue[]) => void;
+}) {
+
+  const propertiesMenu = React.useMemo(() => {
```
[the remark this answers is not in the document]
Comment (swordEdge) [reply src/app/components/CompositionTokenForm.tsx]: fixed

File: src/app/components/EditTokenForm.tsx (hunk 2 above)
Comment (swordEdge) [reply src/app/components/EditTokenForm.tsx]: fixed

**Review by LiamMartens — APPROVED** [review]

**Review by six7 — APPROVED** [review]

**Pull request merged**